pkg/config/spec.go

```diff
@@ -0,0 +1,155 @@
+)
+
+func rename(name string) string {
+	replacements := []struct {
+		old, new string
+	}{
+		{"(R)", ""},
+		{"(r)", ""},
+		{"CPU", ""},
+		{"Processor", ""},
+		{"processor", ""},
+		{"@", ""},
+		{"®", ""},
+		{",", ""},
+	}
+
+	for _, r := range replacements {
+		name = strings.ReplaceAll(name, r.old, r.new)
+	}
+
+	// Remove "n-Bit Multi-Core" pattern
+	re := regexp.MustCompile(`\d+-Bit Multi-Core`)
+	name = re.ReplaceAllString(name, "")
+
+	// Remove anything in parentheses or square brackets
+	re = regexp.MustCompile(`[\(\[].*?[\)\]]`)
+	name = re.ReplaceAllString(name, "")
+
+	// Remove frequency pattern
+	re = regexp.MustCompile(`\d+(\.\d+)?\s?[GMgm][Hh]z`)
+	name = re.ReplaceAllString(name, "")
+
+	return strings.TrimSpace(name)
+}
+
+func formatProcessor(processor string) string {
+	name := rename(processor)
+	name = strings.ToLower(name)
+	// replace spaces and hyphens with an underscore
+	name = strings.ReplaceAll(name, "-", "_")
+	fields := strings.Fields(name)
+	name = strings.Join(fields, "_")
+	return strings.ReplaceAll(name, "_v", "v")
+}
+
+func formatVendor(vendor string) string {
+	vendor = strings.ToLower(vendor)
+	parts := strings.Fields(vendor)
+	if len(parts) > 0 {
+		vendor = parts[0]
+	}
+	vendor = strings.ReplaceAll(vendor, "-", "_")
+	vendor = strings.ReplaceAll(vendor, ",", "")
+	vendor = strings.ReplaceAll(vendor, "'", "")
+	return vendor
+}
+
+func roundToNearestHundred(value float64) int {
+	return int(math.Round(value/100) * 100)
+}
+
+func generateSpec() *MachineSpec {
+	spec := &MachineSpec{}
+	spec.Cores = -1
+	spec.Memory = -1
+	spec.Frequency = -1
+	spec.Chips = 1
+	spec.ThreadsPerCore = 1
+
+	cpus, err := cpu.Info()
+
+	if err == nil {
+		if len(cpus) > 0 {
+			spec.Processor = formatProcessor(cpus[0].ModelName)
+			spec.Vendor = formatVendor(cpus[0].VendorID)
+			spec.Frequency = roundToNearestHundred(cpus[0].Mhz) // rounded frequency
+			physicalIDMap := make(map[string]bool)
+			for index := range cpus {
+				physicalIDMap[cpus[index].PhysicalID] = true
+			}
+			if physicalIDCount := len(physicalIDMap); physicalIDCount > 0 {
+				spec.Chips = physicalIDCount
+			}
+		}
+	}
+
+	if cores, err := cpu.Counts(true); err == nil {
+		spec.Cores = cores
+	}
+	// Threads per core calculation
+	if coresPhysical, err := cpu.Counts(false); err == nil {
+		spec.ThreadsPerCore = spec.Cores / coresPhysical
+	}
+	// Memory info
+	if vmStat, err := mem.VirtualMemory(); err == nil {
+		spec.Memory = int(vmStat.Total / (1024 * 1024 * 1024)) // in GB
+	}
+	return spec
+}
+
+func readMachineSpec(path string) (MachineSpec, error) {
+	specFile, err := os.Open(path)
+	var spec MachineSpec
+	if err == nil {
+		if specBytes, err := io.ReadAll(specFile); err == nil {
+			if err = json.Unmarshal(specBytes, &spec); err == nil {
+				klog.Infof("Read machine spec from config: %v", spec)
+				return spec, nil
+			} else {
+				klog.Infof("failed to unmarshal machine spec %s: %v", string(specBytes), err)
+			}
+		}
+	}
```

Comment:
An idiom in go is to exit early and IMHO, makes the code a bit more readable. 
e.g. this can be reorganised as ... 
Note 
* returning early,
* use of nil
* fmt.Errorf(" ... %w", err) to provide more details to the caller

```
func readMachineSpec(path string) (*MachineSpec, error) {
	specFile, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("failed to open machine spec: %w", err)
	}

	specBytes, err := io.ReadAll(specFile);
	if err != nil {
		return nil, fmt.Errorf("failed to read machine spec: %w", err)
	}

	if err = json.Unmarshal(specBytes, &spec); err != nil {
		return nil, fmt.Errorf("failed to unmarshal machine spec: %v", err")
	}

	klog.V(3).Infof("Read machine spec from config: %v", spec)
	return spec, nil
}
```